### PR TITLE
[core] Pre filter sources when querying by layer

### PR DESCRIPTION
**Pull request by tmpsantos**
Feature querying is extremelly slow because it goes through all the sources and these sources can be dense. This PR makes it possible to filter by source.

Annotations will benefit from it because they live in their own "virtual" source. Also if you have a low density source for say, road disruptions, you can make the query extremely responsive by querying only on this source.

### CODE REVIEW

File: src/mbgl/style/source_impl.cpp [hunk 1]
```diff
@@ -199,6 +199,10 @@ std::unordered_map<std::string, std::vector<Feature>> Source::Impl::queryRendere
         return result;
     }
 
+    if (parameters.sourceIDs && parameters.sourceIDs->find(id) == parameters.sourceIDs->end()) {
```
Comment (brunoabinader) [inline src/mbgl/style/source_impl.cpp]: I'd prefer this check to be done in `Style::queryRenderedFeatures` - this way we can avoid the creation of `result` for sources not present in `sourceIDs`.

**Comment by zugaldia** [comment]
@tmpsantos This is great. cc: @ivovandongen for adapting the Android API.

**Comment by tmpsantos** [comment]
For the record Annotations will get a perf bump for free.

**Comment by tmpsantos** [comment]
> Rather than adding new API surface area, can we optimize this case similarly to mapbox/mapbox-gl-js#2877

Not sure about pre-filtering automatically. Seems like a valid optimization when we don't set the source filter, but I like the filter parameter specially for annotations and the use case that motivated me write this PR in the first place: query features on from a source I created using the dataset editor. I wanted these features to get in sync with native views, so they needed to be queried extremely fast.

Looking at the code, the layers are stored in a `std::vector` and we would have to move it to a `std::set` to get reasonable performance. Is there a reason for having it in a `std::vector` in the first place?

**Comment by 1ec5** [comment]
> query features on from a source I created using the dataset editor

Wouldn’t #5792 be a more direct API for this use case?

**Comment by tmpsantos** [comment]
> Wouldn’t #5792 be a more direct API for this use case?

The low level core API I'm proposing could be used by the platform bindings for implementing this API.

**Comment by 1ec5** [comment]
>> Wouldn’t #5792 be a more direct API for this use case?

> The low level core API I'm proposing could be used by the platform bindings for implementing this API.

This would be a departure from how GL JS implements feature querying: [`querySourceFeatures()`](https://www.mapbox.com/mapbox-gl-js/api/#Map#querySourceFeatures) has quite a few nuances that make it good for the annotation querying use case but not a great candidate for folding into `queryRenderedFeatures()`.

**Comment by tmpsantos** [comment]
@jfirebaugh I re-enabled the benchmarks tests on this PR and wrote a benchmark for querying features. Pre-filtering sources gives a 64x performance bump when the layer you are looking for is in a source not so dense (like annotations or a geojson source).

Explicitly specifying the source makes not measurable difference compared to the pre-filter approach and exposes a new API surface. So let's do how you suggested. :+1:

**Comment by tmpsantos** [comment]
Pre-filtering sources:

```
Benchmark                              Time(ns)    CPU(ns) Iterations
---------------------------------------------------------------------
API_queryRenderedFeatures              14167512   14160000         50                                 
API_queryRenderedFeaturesLayerFilter      21569      21563      33019                                 
```

Bofore the patch:

```
Benchmark                              Time(ns)    CPU(ns) Iterations
---------------------------------------------------------------------
API_queryRenderedFeatures              14428871   14400000         50                                 
API_queryRenderedFeaturesLayerFilter    1309461    1304854        515 
```

**Pull request merged**